sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/parquet/ParquetPartitionReaderFactory.scala

```diff
@@ -0,0 +1,227 @@
+  private val pushDownStringStartWith = sqlConf.parquetFilterPushDownStringStartWith
+  private val pushDownInFilterThreshold = sqlConf.parquetFilterPushDownInFilterThreshold
+
+  override def supportColumnarReads(partition: InputPartition): Boolean = {
+    sqlConf.parquetVectorizedReaderEnabled && sqlConf.wholeStageEnabled &&
+      resultSchema.length <= sqlConf.wholeStageMaxNumFields &&
+      resultSchema.forall(_.dataType.isInstanceOf[AtomicType])
+  }
+
+  override def buildReader(file: PartitionedFile): PartitionReader[InternalRow] = {
+    val reader = if (enableVectorizedReader) {
+      createVectorizedReader(file)
+    } else {
+      createRowBaseReader(file)
+    }
+
+    val fileReader = new PartitionReader[InternalRow] {
+      override def next(): Boolean = reader.nextKeyValue()
+
+      override def get(): InternalRow = reader.getCurrentValue.asInstanceOf[InternalRow]
+
+      override def close(): Unit = reader.close()
+    }
+
+    new PartitionReaderWithPartitionValues(fileReader, readDataSchema,
+      partitionSchema, file.partitionValues)
+  }
+
+  override def buildColumnarReader(file: PartitionedFile): PartitionReader[ColumnarBatch] = {
+    val vectorizedReader = createVectorizedReader(file)
+    vectorizedReader.enableReturningBatches()
+
+    new PartitionReader[ColumnarBatch] {
+      override def next(): Boolean = vectorizedReader.nextKeyValue()
+
+      override def get(): ColumnarBatch =
+        vectorizedReader.getCurrentValue.asInstanceOf[ColumnarBatch]
+
+      override def close(): Unit = vectorizedReader.close()
+    }
+  }
+
+  private def buildReaderBase[T](
+      file: PartitionedFile,
+      buildReaderFunc: (
+        ParquetInputSplit, InternalRow, TaskAttemptContextImpl, Option[FilterPredicate],
+          Option[TimeZone]) => RecordReader[Void, T]): RecordReader[Void, T] = {
+    val conf = broadcastedConf.value.value
+
+    val filePath = new Path(new URI(file.filePath))
+    val split =
+      new org.apache.parquet.hadoop.ParquetInputSplit(
+        filePath,
+        file.start,
+        file.start + file.length,
+        file.length,
+        Array.empty,
+        null)
+
+    lazy val footerFileMetaData =
+      ParquetFileReader.readFooter(conf, filePath, SKIP_ROW_GROUPS).getFileMetaData
+    // Try to push down filters when filter push-down is enabled.
+    val pushed = if (enableParquetFilterPushDown) {
+      val parquetSchema = footerFileMetaData.getSchema
+      val parquetFilters = new ParquetFilters(parquetSchema, pushDownDate, pushDownTimestamp,
+        pushDownDecimal, pushDownStringStartWith, pushDownInFilterThreshold, isCaseSensitive)
+      filters
+        // Collects all converted Parquet filter predicates. Notice that not all predicates can be
+        // converted (`ParquetFilters.createFilter` returns an `Option`). That's why a `flatMap`
+        // is used here.
+        .flatMap(parquetFilters.createFilter)
+        .reduceOption(FilterApi.and)
```

Comment:
`ParquetScanBuilder` already did most thing. Here, I guess we need `.reduceOption(FilterApi.and)` only.

Reply:
We still need to do the actual push down here.
See my comments above: https://github.com/apache/spark/pull/24327#discussion_r289266681

Reply:
`actual` pushdown implies the other pushdowns in DSv2 is `not-actual`. Also, this seems to make `pushedFilters` in the explained plan inconsistent with the one Parquet library received. I believe this will cause lots of confusion.

Reply:
We need the filter push down here as per the discussion in https://github.com/apache/spark/pull/21696#pullrequestreview-133842990 .